pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -1315,6 +1330,82 @@ func (m *kubeGenericRuntimeManager) SyncPod(ctx context.Context, pod *v1.Pod, po
 	return
 }
 
+// imageVolumePulls are the pull results for each image volume name.
+type imageVolumePulls = map[string]imageVolumePullResult
+
+// imageVolumePullResult is a pull result for a single image volume.
+// If spec is nil, then err and msg should be set.
+// If err is nil, then spec should be set.
+type imageVolumePullResult struct {
+	spec runtimeapi.ImageSpec
+	err  error
+	msg  string
+}
+
+func (m *kubeGenericRuntimeManager) toKubeContainerImageVolumes(imageVolumePullResults imageVolumePulls, container *v1.Container, pod *v1.Pod, syncResult *kubecontainer.SyncResult) (kubecontainer.ImageVolumes, error) {
+	if len(imageVolumePullResults) == 0 {
+		return nil, nil
+	}
+
+	imageVolumes := kubecontainer.ImageVolumes{}
+	for _, v := range container.VolumeMounts {
+		res, ok := imageVolumePullResults[v.Name]
+		if !ok {
+			continue
+		}
+
+		if res.err != nil {
+			s, _ := grpcstatus.FromError(res.err)
+			m.recordContainerEvent(pod, container, "", v1.EventTypeWarning, events.FailedToCreateContainer, "Error: %v", s.Message())
+			syncResult.Fail(res.err, res.msg)
+			return nil, res.err
```

Comment:
Perhaps we should aggregate all the errors from all of the container.VolumeMounts.

Reply:
Alright, we now throw events for all failed pulls per container, but we can just return one syncresult error. I've chosen to report the last one, which makes the implementation a bit simpler.